app/queries/draft_and_live_versions.rb

```diff
@@ -0,0 +1,56 @@
+module Queries
+  module DraftAndLiveVersions
+    extend ArelHelpers
+    def self.call(content_item, target_class, locale = nil, base_path = nil)
+      content_items_table = table(:content_items)
+      states_table = table(:states)
+      translations_table = table(:translations)
+      locations_table = table(:locations)
+      versions_table = table(target_class)
+      versions_fk = target_class == "lock_versions" ? :target_id : :content_item_id
+
+      if base_path.nil?
+        base_path = locations_table.project(locations_table[:base_path])
+          .where(locations_table[:content_item_id].eq(content_item.id))
+      end
+      if locale.nil?
+        locale = translations_table.project(translations_table[:locale])
+          .where(translations_table[:content_item_id].eq(content_item.id))
+      end
+
+      scope = content_items_table
+        .project(
+          Arel::Nodes::NamedFunction.new("row_number", [])
+            .over(Arel::Nodes::Window.new
+              .partition(states_table[:name])
+              .order(versions_table[:number])
+            ).as("r"),
+          states_table[:name].as("state"),
+          versions_table[:number].as("version_number")
+        )
+        .outer_join(states_table).on(
+          content_items_table[:id].eq(states_table[:content_item_id])
+        )
+        .outer_join(translations_table).on(
+          content_items_table[:id].eq(translations_table[:content_item_id])
+        )
+        .outer_join(locations_table).on(
+          content_items_table[:id].eq(locations_table[:content_item_id])
+        )
+        .outer_join(versions_table).on(
+          content_items_table[:id].eq(versions_table[versions_fk])
+        )
+        .where(content_items_table[:content_id].eq(content_item.content_id))
+        .where(locations_table[:base_path].eq(base_path))
+        .where(translations_table[:locale].eq(locale))
+        .where(states_table[:name].in(%w(draft published)))
+
+      partitioned = cte(scope, as: :partitioned)
+      query = partitioned.table.project(Arel.star)
+        .with(partitioned.compiled_scope)
+        .where(partitioned.table[:r].eq(1))
+
+      Hash[get_rows(query).map { |i| [i["state"], i["version_number"].to_i] }]
```

Comment:
And using a CTE makes this a little neater, so now I have something that looks like:

```
EXPLAIN ANALYZE WITH content_store_and_lock_versions AS (
    SELECT
        CASE WHEN states.name IN ('published', 'unpublished') THEN 'live'
             WHEN states.name = 'draft' THEN 'draft' END
        AS content_store,
        lock_versions.number AS lock_version
    FROM content_items
    INNER JOIN states ON content_items.id = states.content_item_id
    INNER JOIN locations ON content_items.id = locations.content_item_id
    INNER JOIN translations ON content_items.id = translations.content_item_id
    INNER JOIN lock_versions ON content_items.id = lock_versions.target_id
        AND lock_versions.target_type = 'ContentItem'
    WHERE content_items.content_id = '...'
        AND states.name IN ('published', 'unpublished', 'draft')
)
(
    SELECT content_store, lock_version
    FROM content_store_and_lock_versions
    WHERE content_store = 'live'
    ORDER BY lock_version DESC
    LIMIT 1
)
UNION
(
    SELECT content_store, lock_version
    FROM content_store_and_lock_versions
    WHERE content_store = 'draft'
    ORDER BY lock_version DESC
    LIMIT 1
);
```

Not sure exactly what the conditions on the content items need to be, but its all in one place now.